diff --git a/src/config/FreeRTOSConfig.h b/src/config/FreeRTOSConfig.h
index ad9af7c2..aebfa073 100644
--- a/src/config/FreeRTOSConfig.h
+++ b/src/config/FreeRTOSConfig.h
@@ -74,8 +74,10 @@
 #define configUSE_IDLE_HOOK			1
 #define configUSE_TICK_HOOK			0
 #define configCPU_CLOCK_HZ			( ( unsigned long ) FREERTOS_MCU_CLOCK_HZ )
-#define configTICK_RATE_HZ_RAW  1000
+#define configTICK_RATE_HZ_RAW  800
 #define configTICK_RATE_HZ			( ( portTickType ) configTICK_RATE_HZ_RAW )
+#define configTICK_RATE_HZ_RAW_BROKEN  1000
+#define configTICK_RATE_HZ_BROKEN      ( ( portTickType ) configTICK_RATE_HZ_RAW_BROKEN )
 #define configMINIMAL_STACK_SIZE	( ( unsigned short ) FREERTOS_MIN_STACK_SIZE )
 #define configTOTAL_HEAP_SIZE		( ( size_t ) ( FREERTOS_HEAP_SIZE ) )
 #define configMAX_TASK_NAME_LEN		( 10 )
@@ -121,16 +123,16 @@ to exclude the API function. */
 #define vPortSVCHandler SVC_Handler
 
 //Milliseconds to OS Ticks
-#if configTICK_RATE_HZ_RAW != 1000
+#if configTICK_RATE_HZ_RAW_BROKEN != 1000
   #error "Please review the use of M2T and T2M if there is not a 1 to 1 mapping between ticks and milliseconds"
 #endif
 #define M2T(X) ((unsigned int)(X))
-#define F2T(X) ((unsigned int)((configTICK_RATE_HZ/(X))))
+#define F2T(X) ((unsigned int)((configTICK_RATE_HZ_BROKEN/(X))))
 #define T2M(X) ((unsigned int)(X))
 
 // Seconds to OS ticks
-#define S2T(X) ((portTickType)((X) * configTICK_RATE_HZ))
-#define T2S(X) ((X) / (float)configTICK_RATE_HZ)
+#define S2T(X) ((portTickType)((X) * configTICK_RATE_HZ_BROKEN))
+#define T2S(X) ((X) / (float)configTICK_RATE_HZ_BROKEN)
 
 
 // DEBUG SECTION
